src/rp2_common/hardware_gpio/include/hardware/gpio.h

```diff
@@ -365,46 +375,243 @@ enum gpio_drive_strength gpio_get_drive_strength(uint gpio);
+ * it is possible to add explicit GPIO IRQ handlers which are called independent of the default event callback.
+ *
+ * This method adds such a callback, and disables the "default" callback for the specified GPIOs.
+ *
+ * A raw handler should check for whichever GPIOs and events it handles, and acknowledge them itself; it might look something like:
+ *
+ * \code{.c}
+ * void my_irq_handler(void) {
+ *     if (gpio_get_irq_events(my_gpio_num) & my_gpio_events) {
+ *        gpio_acknowledge_irq(my_gpio_num, my_gpio_events);
+ *       // handle the IRQ
+ *     }
+ *     if (gpio_get_irq_events(my_gpio_num2) & my_gpio_events2) {
+ *        gpio_acknowledge_irq(my_gpio_num2, my_gpio_events2);
+ *       // handle the IRQ
+ *     }
+ * }
+ * \code
+ *
+ * @param gpio_mask a bit mask of the GPIO numbers that will no longer be passed to the default callback for this core
+ * @param handler the handler to add to the list of GPIO IRQ handlers for this core
+ */
+void gpio_add_raw_irq_handler_masked(uint gpio_mask, irq_handler_t handler);
+
+/*! \brief Adds a raw GPIO IRQ handler for a specific GPIO on the current code
+ *  \ingroup hardware_gpio
+ *
+ * In addition to the default mechanism of a single GPIO IRQ event callback per core (see \ref gpio_set_irq_callback),
+ * it is possible to add explicit GPIO IRQ handlers which are called independent of the default event callback.
+ *
+ * This method adds such a callback, and disables the "default" callback for the specified GPIO.
+ *
+ * A raw handler should check for whichever GPIOs and events it handles, and acknowledge them itself; it might look something like:
+ *
+ * \code{.c}
+ * void my_irq_handler(void) {
+ *     if (gpio_get_irq_events(my_gpio_num) & my_gpio_events) {
+ *        gpio_acknowledge_irq(my_gpio_num, my_gpio_events);
+ *       // handle the IRQ
+ *     }
+ * }
+ * \code
+ *
+ * @param gpio the GPIO number that will no longer be passed to the default callback for this core
+ * @param handler the handler to add to the list of GPIO IRQ handlers for this core
+ */
+static inline void gpio_add_raw_irq_handler(uint gpio, irq_handler_t handler) {
+    check_gpio_param(gpio);
+    gpio_add_raw_irq_handler_masked(1u << gpio, handler);
+}
+
+
+/*! \brief Removes a raw GPIO IRQ handler for the specified GPIOs on the current core
+ *  \ingroup hardware_gpio
+ *
+ * In addition to the default mechanism of a single GPIO IRQ event callback per core (see \ref gpio_set_irq_callback),
+ * it is possible to add explicit GPIO IRQ handlers which are called independent of the default event callback.
+ *
+ * This method removes such a callback, and enables the "default" callback for the specified GPIOs.
+ *
+ * @param gpio_mask a bit mask of the GPIO numbers that will now be passed to the default callback for this core
+ * @param handler the handler to remove from the list of GPIO IRQ handlers for this core
```

Comment:
If you use `gpio_add_raw_irq_handler_masked` to e.g. set up GPIOs 2, 3 and 4 to be handled by `my_callback1`, is it possible that you might want to switch GPIOs 2 and 4 back to the default handler, but **not** remove the `my_callback1` handler that's still being used by GPIO 3?

Reply:
Good point! I think this code works fine as long as you add/remove with the same masks, and some cases where you fail to do this will be caught by an assert on removing the same IRQ twice, but in this case you would actually just lock up the core because nobody would clear the IRQ status.

I don't see any way to fix this without excessive bookkeeping, so maybe we just need to document that add/remove must use the same mask?

Reply:
fixed